lib/absinthe/phase/document/complexity/analysis.ex

```diff
@@ -0,0 +1,76 @@
+defmodule Absinthe.Phase.Document.Complexity.Analysis do
+  @moduledoc false
+
+  # Analyses document complexity.
+
+  alias Absinthe.{Blueprint, Phase, Complexity}
+
+  use Absinthe.Phase
+
+  @default_complexity 1
+
+  @doc """
+  Run complexity analysis.
+  """
+  @spec run(Blueprint.t, Keyword.t) :: Phase.result_t
+  def run(input, options \\ []) do
+    info_data = info_boilerplate(input, options)
+    fun = &handle_node(&1, info_data)
+    {:ok, Blueprint.update_current(input, &Blueprint.postwalk(&1, fun))}
+  end
+
+  def handle_node(%Blueprint.Document.Field{complexity: nil,
+                                            fields: fields,
+                                            argument_data: args,
+                                            schema_node: schema_node} = node, info_data) do
+    child_complexity = sum_complexity(fields)
+    complexity = field_complexity(schema_node, args, child_complexity, info_data, node)
+    %{node | complexity: complexity}
+  end
+  def handle_node(%Blueprint.Document.Operation{complexity: nil, fields: fields} = node, _) do
+    complexity = sum_complexity(fields)
+    %{node | complexity: complexity}
+  end
+  def handle_node(node, _) do
+    node
+  end
+
+  defp field_complexity(%{complexity: nil}, _, child_complexity, _, _) do
+    @default_complexity + child_complexity
+  end
+  defp field_complexity(%{complexity: complexity}, _, _, _, _)
+       when is_integer(complexity) and complexity >= 0 do
+    complexity
+  end
+  defp field_complexity(%{complexity: complexity}, arg, child_complexity, _, _)
+       when is_function(complexity, 2) do
+    complexity.(arg, child_complexity)
+  end
+  defp field_complexity(%{complexity: complexity}, arg, child_complexity, info_data, node)
+       when is_function(complexity, 3) do
+    info = struct(Complexity, Map.put(info_data, :definition, node))
+    complexity.(arg, child_complexity, info)
+  end
+  defp field_complexity(%{complexity: {mod, fun}}, arg, child_complexity, info_data, _) do
+    info = struct(Complexity, Map.put(info_data, :definition, node))
+    apply(mod, fun, [arg, child_complexity, info])
```

Comment:
Example situation where `max` may be useful: `{foo(limit: -1000) {y], x { ...ThisCanBeArbitrarilyComplexNow}`. If the logic for computing the complexity of `foo` is not careful to restrict the values of `limit` the `foo` field would have substantial negative complexity, which gives wide leeway for the rest of the document to be as complex as you want.

While obviously complexity functions ought to take precautions, I feel like preventing this kind of attack is within the purview of what this phase should do.